File: backend/nlp_service.py
"""
spaCy NLP Service for TimeTracker
Advanced natural language processing for legal time entries
"""

import spacy
from flask import Flask, request, jsonify
from flask_cors import CORS
import re
from datetime import datetime
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = Flask(__name__)
CORS(app)  # Enable CORS for frontend communication

# Load spaCy model
try:
    nlp = spacy.load("en_core_web_lg")
    logger.info("spaCy model loaded successfully")
except Exception as e:
    logger.error(f"Error loading spaCy model: {e}")
    nlp = None

# Legal practice areas and keywords
LEGAL_PRACTICE_AREAS = {
    'corporate': ['incorporation', 'merger', 'acquisition', 'compliance', 'corporate'],
    'litigation': ['lawsuit', 'trial', 'pleading', 'motion', 'discovery', 'deposition'],
    'contract': ['contract', 'agreement', 'negotiation', 'review', 'drafting'],
    'real_estate': ['property', 'lease', 'closing', 'title', 'mortgage', 'real estate'],
    'family': ['divorce', 'custody', 'separation', 'support', 'family law'],
    'criminal': ['criminal', 'defense', 'prosecution', 'arraignment', 'bail'],
    'employment': ['employment', 'termination', 'discrimination', 'harassment'],
    'ip': ['patent', 'trademark', 'copyright', 'intellectual property', 'licensing']
}

# Time patterns for extraction
TIME_PATTERNS = [
    (r'(\d+(?:\.\d+)?)\s*(?:hours?|hrs?|h)', 'hours'),
    (r'(\d+)\s*(?:minutes?|mins?|m)', 'minutes'),
    (r'(\d+):(\d+)', 'clock'),  # 2:30 format
]

def extract_time(text):
    """Extract time duration from text"""
    hours = 0
    minutes = 0
    confidence = 0
    
    text_lower = text.lower()
    
    # Try each pattern
    for pattern, time_type in TIME_PATTERNS:
        match = re.search(pattern, text_lower)
        if match:
            if time_type == 'hours':
                hours = float(match.group(1))
                confidence = 95
            elif time_type == 'minutes':
                minutes = int(match.group(1))
                confidence = 95
            elif time_type == 'clock':
                hours = int(match.group(1))
                minutes = int(match.group(2))
                confidence = 90
    
    # Convert decimal hours to hours and minutes
    if hours > 0 and hours % 1 != 0:
        minutes += int((hours % 1) * 60)
        hours = int(hours)
    
    return {
        'hours': int(hours),
        'minutes': int(minutes),
        'confidence': confidence
    }

def detect_practice_area(text):
    """Detect legal practice area from text"""
    text_lower = text.lower()
    
    for area, keywords in LEGAL_PRACTICE_AREAS.items():
        for keyword in keywords:
            if keyword in text_lower:
                return area
    
    return 'general'

def extract_entities(text):
    """Extract named entities using spaCy"""
    if not nlp:
        return {
            'persons': [],
            'organizations': [],
            'tasks': [],
            'confidence': 0
        }
    
    doc = nlp(text)
    
    entities = {
        'persons': [],
        'organizations': [],
        'tasks': [],
        'dates': [],
        'money': []
    }
    
    for ent in doc.ents:
        if ent.label_ == 'PERSON':
            entities['persons'].append(ent.text)
        elif ent.label_ == 'ORG':
            entities['organizations'].append(ent.text)
        elif ent.label_ == 'DATE':
            entities['dates'].append(ent.text)
        elif ent.label_ == 'MONEY':
            entities['money'].append(ent.text)
    
    # Extract tasks (verb phrases)
    for chunk in doc.noun_chunks:
        # Look for action verbs preceding the chunk
        if chunk.root.head.pos_ == 'VERB':
            task = f"{chunk.root.head.text} {chunk.text}"
            entities['tasks'].append(task)
    
    return entities

@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({
        'status': 'healthy',
        'spacy_loaded': nlp is not None,
        'timestamp': datetime.now().isoformat()
    })

@app.route('/parse-time-entry', methods=['POST'])
def parse_time_entry():
    """Parse natural language time entry"""
    try:
        data = request.json
        text = data.get('text', '')
        
        if not text:
            return jsonify({'error': 'No text provided'}), 400
        
        # Extract time
        time_data = extract_time(text)
        
        # Extract entities
        entities = extract_entities(text)
        
        # Detect practice area
        practice_area = detect_practice_area(text)
        
        # Determine client (prefer organization over person)
        client = ''
        if entities['organizations']:
            client = entities['organizations'][0]
        elif entities['persons']:
            client = entities['persons'][0]
        
        # Determine matter/task
        matter = ''
        if entities['tasks']:
            matter = entities['tasks'][0]
        
        # Calculate overall confidence
        confidence = 0
        confidence_count = 0
        
        if time_data['confidence'] > 0:
            confidence += time_data['confidence']
            confidence_count += 1
        
        if client:
            confidence += 90
            confidence_count += 1
        
        if matter:
            confidence += 80
            confidence_count += 1
        
        overall_confidence = confidence / confidence_count if confidence_count > 0 else 0
        
        result = {
            'client': client,
            'matter': matter,
            'hours': time_data['hours'],
            'minutes': time_data['minutes'],
            'practiceArea': practice_area,
            'description': text,
            'confidence': int(overall_confidence),
            'entities': entities,
            'metadata': {
                'timeConfidence': time_data['confidence'],
                'source': 'spacy'
            }
        }
        
        logger.info(f"Parsed time entry: {result}")
        return jsonify(result)
        
    except Exception as e:
        logger.error(f"Error parsing time entry: {e}")
        return jsonify({'error': str(e)}), 500

@app.route('/extract-entities', methods=['POST'])
def extract_entities_endpoint():
    """Extract entities from text"""
    try:
        data = request.json
        text = data.get('text', '')
        
        if not text:
            return jsonify({'error': 'No text provided'}), 400
        
        entities = extract_entities(text)
        practice_area = detect_practice_area(text)
        
        result = {
            'entities': entities,
            'practiceArea': practice_area,
            'confidence': 85 if entities['organizations'] or entities['persons'] else 50
        }
        
        return jsonify(result)
        
    except Exception as e:
        logger.error(f"Error extracting entities: {e}")
        return jsonify({'error': str(e)}), 500

@app.route('/analyze-legal-text', methods=['POST'])
def analyze_legal_text():
    """Analyze legal text for key concepts"""
    try:
        data = request.json
        text = data.get('text', '')
        
        if not text:
            return jsonify({'error': 'No text provided'}), 400
        
        if not nlp:
            return jsonify({'error': 'spaCy model not loaded'}), 500
        
        doc = nlp(text)
        
        # Extract key information
        analysis = {
            'entities': extract_entities(text),
            'practice_area': detect_practice_area(text),
            'sentences': [sent.text for sent in doc.sents],
            'key_phrases': [chunk.text for chunk in doc.noun_chunks],
            'complexity': len(list(doc.sents))  # Simple complexity metric
        }
        
        return jsonify(analysis)
        
    except Exception as e:
        logger.error(f"Error analyzing legal text: {e}")
        return jsonify({'error': str(e)}), 500

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=True)

File: services/nlpService.js
// Natural Language Processing Service
// Enhanced NLP/NLU capabilities for time entry parsing and entity recognition
// Uses backend spaCy service for advanced NLP processing

/**
 * Enhanced NLP service for time tracking entries
 * Provides entity recognition, intent detection, and context extraction
 * Now uses backend spaCy service for better accuracy
 */

// Backend API URL - default to localhost for development
const BACKEND_URL = import.meta.env.VITE_BACKEND_URL || 'http://localhost:5000'

// Common legal practice areas and tasks
const LEGAL_TASKS = {
  corporate: ['drafting', 'review', 'amendment', 'incorporation', 'merger', 'acquisition', 'due diligence', 'compliance'],
  litigation: ['pleading', 'motion', 'discovery', 'deposition', 'trial prep', 'court appearance', 'research'],
  contract: ['negotiation', 'review', 'drafting', 'amendment', 'execution'],
  real_estate: ['closing', 'title search', 'lease review', 'purchase agreement', 'mortgage'],
  family: ['custody', 'divorce', 'separation', 'support', 'mediation'],
  criminal: ['arraignment', 'bail', 'plea', 'sentencing', 'appeal'],
  employment: ['termination', 'discrimination', 'harassment', 'contract review'],
  ip: ['patent', 'trademark', 'copyright', 'licensing', 'infringement']
}

// Time indicators
const TIME_PATTERNS = {
  hours: /(\d+(?:\.\d+)?)\s*(?:hours?|hrs?|h)/i,
  minutes: /(\d+)\s*(?:minutes?|mins?|m)/i,
  decimal: /(\d+\.\d+)\s*(?:hours?)?/i,
  fraction: /(\d+)\s*(?:and\s+)?(\d+)\/(\d+)\s*(?:hours?)?/i,
  range: /(\d+)-(\d+)\s*(?:hours?|hrs?)/i
}

// Client name patterns
const CLIENT_PATTERNS = {
  for: /\b(?:for|client)\s+([A-Z][A-Za-z\s&.,']+?)(?:\s+on|\s+re:|\s+-|\s+about|$)/i,
  with: /\b(?:with|meeting\s+with)\s+([A-Z][A-Za-z\s&.,']+?)(?:\s+on|\s+re:|\s+about|$)/i,
  company: /\b([A-Z][A-Za-z\s]+(?:Inc|LLC|Ltd|Corp|Company|LLP|PC)\.?)/i,
  name: /\b([A-Z][a-z]+\s+[A-Z][a-z]+)/
}

// Matter/task patterns
const MATTER_PATTERNS = {
  re: /\b(?:re:|regarding|about|on)\s+(.+?)(?:\s+for|\s+with|\s+\d+|$)/i,
  gerund: /\b(drafting|reviewing|preparing|analyzing|researching|attending|conducting|filing|negotiating|revising)\s+(.+?)(?:\s+for|\s+\d+|$)/i,
  noun: /\b(?:draft|review|prepare|analyze|research|attend|conduct|file|negotiate|revise)\s+(.+?)(?:\s+for|\s+\d+|$)/i
}

/**
 * Call backend spaCy service for enhanced NLP processing
 */
const callBackendNLP = async (text) => {
  try {
    const response = await fetch(`${BACKEND_URL}/parse-time-entry`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ text })
    })

    if (!response.ok) {
      throw new Error(`Backend service error: ${response.status}`)
    }

    const result = await response.json()
    return result
  } catch (error) {
    console.warn('Backend spaCy service unavailable, using local fallback:', error)
    return null
  }
}

/**
 * Enhance task description using backend spaCy service
 */
export const enhanceTaskDescription = async (description, clientType, matterType) => {
  try {
    // For now, use local enhancement until backend has this endpoint
    // This provides immediate feedback while maintaining functionality
    return enhanceDescriptionLocal(description, clientType, matterType)
  } catch (error) {
    console.error('Error enhancing description:', error)
    return description
  }
}

/**
 * Local description enhancement (fallback)
 */
const enhanceDescriptionLocal = (description, clientType, matterType) => {
  if (!description || description.trim().length === 0) {
    return description
  }

  let enhanced = description.trim()

  // Capitalize first letter
  enhanced = enhanced.charAt(0).toUpperCase() + enhanced.slice(1)

  // Add period if missing
  if (!enhanced.match(/[.!?]$/)) {
    enhanced += '.'
  }

  // Common legal task enhancements
  const enhancements = {
    'call': 'Telephone conference',
    'email': 'Email correspondence',
    'meeting': 'Client meeting',
    'review': 'Reviewed and analyzed',
    'draft': 'Drafted',
    'research': 'Legal research regarding',
    'prepare': 'Prepared',
    'revise': 'Revised',
    'attend': 'Attended'
  }

  // Apply enhancements for common short descriptions
  for (const [key, value] of Object.entries(enhancements)) {
    const regex = new RegExp(`^${key}\\b`, 'i')
    if (regex.test(enhanced)) {
      enhanced = enhanced.replace(regex, value)
      break
    }
  }

  return enhanced
}

/**
 * Extract time duration from natural language text
 */
export const extractTime = (text) => {
  let hours = 0
  let minutes = 0

  // Check for decimal hours (e.g., "2.5 hours")
  const decimalMatch = text.match(TIME_PATTERNS.decimal)
  if (decimalMatch) {
    const decimal = parseFloat(decimalMatch[1])
    hours = Math.floor(decimal)
    minutes = Math.round((decimal - hours) * 60)
    return { hours, minutes, confidence: 0.95 }
  }

  // Check for explicit hours
  const hoursMatch = text.match(TIME_PATTERNS.hours)
  if (hoursMatch) {
    hours = parseFloat(hoursMatch[1])
  }

  // Check for explicit minutes
  const minutesMatch = text.match(TIME_PATTERNS.minutes)
  if (minutesMatch) {
    minutes = parseInt(minutesMatch[1])
  }

  // Check for fraction (e.g., "1 and 1/2 hours")
  const fractionMatch = text.match(TIME_PATTERNS.fraction)
  if (fractionMatch) {
    hours = parseInt(fractionMatch[1])
    const numerator = parseInt(fractionMatch[2])
    const denominator = parseInt(fractionMatch[3])
    minutes = Math.round((numerator / denominator) * 60)
    return { hours, minutes, confidence: 0.9 }
  }

  // Check for range (take average)
  const rangeMatch = text.match(TIME_PATTERNS.range)
  if (rangeMatch) {
    const avg = (parseInt(rangeMatch[1]) + parseInt(rangeMatch[2])) / 2
    hours = Math.floor(avg)
    minutes = Math.round((avg - hours) * 60)
    return { hours, minutes, confidence: 0.7 }
  }

  // Default confidence based on whether we found time info
  const confidence = (hours > 0 || minutes > 0) ? 0.85 : 0.0

  return { hours, minutes, confidence }
}

/**
 * Extract client name from text with improved accuracy
 */
export const extractClient = (text) => {
  // Try "for" pattern first (highest confidence)
  let match = text.match(CLIENT_PATTERNS.for)
  if (match) {
    const client = match[1].trim()
    // Clean up common trailing words
    const cleaned = client.replace(/\s+(on|re|about|regarding)$/i, '').trim()
    return { client: cleaned, confidence: 0.9 }
  }

  // Try "with" pattern
  match = text.match(CLIENT_PATTERNS.with)
  if (match) {
    const client = match[1].trim()
    const cleaned = client.replace(/\s+(on|re|about|regarding)$/i, '').trim()
    return { client: cleaned, confidence: 0.85 }
  }

  // Try company pattern (Inc, LLC, Corp, etc.)
  match = text.match(CLIENT_PATTERNS.company)
  if (match) {
    return { client: match[1].trim(), confidence: 0.8 }
  }

  // Try person name pattern (First Last)
  match = text.match(CLIENT_PATTERNS.name)
  if (match) {
    const name = match[1].trim()
    // Only accept if it's a reasonable length (2-50 chars)
    if (name.length >= 2 && name.length <= 50) {
      return { client: name, confidence: 0.6 }
    }
  }

  return { client: '', confidence: 0.0 }
}

/**
 * Extract matter/task description from text with improved accuracy
 */
export const extractMatter = (text) => {
  // Try "re:" pattern first (most explicit)
  let match = text.match(MATTER_PATTERNS.re)
  if (match) {
    const matter = match[1].trim()
    // Clean up matter description - remove trailing prepositions
    const cleaned = matter.replace(/\s+(for|with)$/i, '').trim()
    return { matter: cleaned, confidence: 0.9 }
  }

  // Try gerund pattern (e.g., "drafting contract", "reviewing agreement")
  match = text.match(MATTER_PATTERNS.gerund)
  if (match) {
    const matter = `${match[1]} ${match[2]}`.trim()
    // Capitalize first letter
    const formatted = matter.charAt(0).toUpperCase() + matter.slice(1)
    return { matter: formatted, confidence: 0.85 }
  }

  // Try noun pattern (e.g., "draft contract", "review agreement")
  match = text.match(MATTER_PATTERNS.noun)
  if (match) {
    const matter = match[1].trim()
    const cleaned = matter.replace(/\s+(for|with)$/i, '').trim()
    const formatted = cleaned.charAt(0).toUpperCase() + cleaned.slice(1)
    return { matter: formatted, confidence: 0.75 }
  }

  return { matter: '', confidence: 0.0 }
}

/**
 * Detect practice area from text
 */
export const detectPracticeArea = (text) => {
  const lowerText = text.toLowerCase()
  let bestMatch = { area: 'general', confidence: 0.0 }

  for (const [area, keywords] of Object.entries(LEGAL_TASKS)) {
    for (const keyword of keywords) {
      if (lowerText.includes(keyword)) {
        const confidence = 0.7 + (keyword.length / 100) // Longer keywords = higher confidence
        if (confidence > bestMatch.confidence) {
          bestMatch = { area, confidence: Math.min(confidence, 0.95) }
        }
      }
    }
  }

  return bestMatch
}

/**
 * Extract entities from natural language time entry with improved accuracy
 * First tries backend spaCy service, falls back to local processing
 * Returns: { time, client, matter, description, practiceArea, confidence }
 */
export const parseTimeEntry = async (text) => {
  if (!text || text.trim().length === 0) {
    return null
  }

  // Try backend spaCy service first
  const backendResult = await callBackendNLP(text)
  if (backendResult && backendResult.confidence >= 70) {
    return {
      hours: backendResult.hours || 0,
      minutes: backendResult.minutes || 0,
      client: backendResult.client || '',
      matter: backendResult.matter || '',
      description: backendResult.description || text,
      practiceArea: backendResult.practiceArea || 'general',
      confidence: backendResult.confidence,
      source: 'spacy-backend'
    }
  }

  // Fallback to local processing
  const time = extractTime(text)
  const client = extractClient(text)
  const matter = extractMatter(text)
  const practiceArea = detectPracticeArea(text)

  // Generate a clean description by removing time indicators and common patterns
  let description = text
    .replace(TIME_PATTERNS.hours, '')
    .replace(TIME_PATTERNS.minutes, '')
    .replace(TIME_PATTERNS.decimal, '')
    .replace(/\blog\b/gi, '') // Remove "log" keyword
    .replace(/\s+/g, ' ')
    .trim()

  // If we have a matter description, use it as the description
  if (matter.matter && matter.confidence >= 0.7) {
    description = matter.matter
  }

  // Clean up description - remove client name and common words
  if (client.client && description.toLowerCase().includes(client.client.toLowerCase())) {
    description = description.replace(new RegExp(client.client, 'gi'), '').trim()
  }
  description = description
    .replace(/\b(for|with|client|re:|regarding|about)\b/gi, '')
    .replace(/\s+/g, ' ')
    .trim()

  // Capitalize first letter
  if (description.length > 0) {
    description = description.charAt(0).toUpperCase() + description.slice(1)
  }

  // Calculate overall confidence with improved weighting
  const confidence = (
    time.confidence * 0.35 +
    client.confidence * 0.35 +
    matter.confidence * 0.20 +
    practiceArea.confidence * 0.10
  )

  return {
    hours: time.hours,
    minutes: time.minutes,
    client: client.client,
    matter: matter.matter || description || 'General matter',
    description: description || text,
    practiceArea: practiceArea.area,
    confidence: Math.round(confidence * 100),
    source: 'local-fallback',
    metadata: {
      timeConfidence: Math.round(time.confidence * 100),
      clientConfidence: Math.round(client.confidence * 100),
      matterConfidence: Math.round(matter.confidence * 100),
      practiceAreaConfidence: Math.round(practiceArea.confidence * 100)
    }
  }
}

/**
 * Suggest similar clients based on input
 */
export const suggestClients = (input, existingClients) => {
  if (!input || input.length < 2) return []

  const lowerInput = input.toLowerCase()
  return existingClients
    .filter(client => 
      client.name.toLowerCase().includes(lowerInput) ||
      (client.company && client.company.toLowerCase().includes(lowerInput))
    )
    .slice(0, 5)
}

/**
 * Suggest task descriptions based on practice area and context
 */
export const suggestTasks = (practiceArea, context = '') => {
  const tasks = LEGAL_TASKS[practiceArea] || LEGAL_TASKS.corporate
  const lowerContext = context.toLowerCase()

  // Filter tasks relevant to context
  let relevantTasks = tasks
  if (context.length > 2) {
    relevantTasks = tasks.filter(task => 
      task.includes(lowerContext) || lowerContext.includes(task)
    )
  }

  // If no relevant tasks, return all tasks for the practice area
  if (relevantTasks.length === 0) {
    relevantTasks = tasks
  }

  // Format as proper sentences
  return relevantTasks.slice(0, 10).map(task => {
    const words = task.split(' ')
    return words.map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ')
  })
}

/**
 * Validate and normalize time entry data
 */
export const validateTimeEntry = (entry) => {
  const errors = []
  const warnings = []

  // Check required fields
  if (!entry.client || entry.client.trim().length === 0) {
    errors.push('Client name is required')
  }

  if (!entry.description || entry.description.trim().length === 0) {
    errors.push('Task description is required')
  }

  if (entry.hours === 0 && entry.minutes === 0) {
    errors.push('Time duration must be greater than 0')
  }

  // Check for unusual values
  if (entry.hours > 24) {
    warnings.push('Time exceeds 24 hours - please verify')
  }

  if (entry.minutes > 0 && entry.minutes % 6 !== 0) {
    warnings.push('Minutes not in 6-minute increments (standard billing practice)')
  }

  return {
    isValid: errors.length === 0,
    errors,
    warnings
  }
}

/**
 * Format time entry for display
 */
export const formatTimeEntry = (entry) => {
  const { hours, minutes } = entry
  const totalHours = hours + (minutes / 60)
  
  return {
    displayTime: `${hours}h ${minutes}m`,
    decimalTime: totalHours.toFixed(2),
    billingUnits: Math.ceil(totalHours * 10) / 10 // Round to nearest 0.1
  }
}

// Export all functions as default for backward compatibility
export default {
  extractTime,
  extractClient,
  extractMatter,
  detectPracticeArea,
  parseTimeEntry,
  suggestClients,
  suggestTasks,
  validateTimeEntry,
  formatTimeEntry
}
